### ISSUE TITLE
bad suggestion for Redundant $

### ISSUE THREAD

**Issue by JasonMFry** [ndmitchell/hlint#1322]
Not sure if this is helpful or not, but this might be a bug. I'm not sure if it's known or unknown, I didn't see any relevant issues in my brief searching.

relevant code
```haskell
import UnliftIO (withRunInIO)

main :: IO ()
main = do
  config@Config{..} <- Config.fromEnvironmentVariables
  withEnvironment config $ \env -> do
    App.runApp env $
      withRunInIO $ \runInIO ->
        runInIO $ startApi configHttp

withEnvironment :: Config -> (App.Environment -> IO a) -> IO a
withEnvironment Config{..} useEnvironment = -- irrelevant to the issue

startApi :: HttpConfig -> App ()
startApi HttpConfig{..} = -- irrelevant to the issue

```

hlint suggestion
```
Found:
  App.runApp env
    $ withRunInIO $ \ runInIO -> runInIO $ startApi configHttp
Perhaps:
  App.runApp env . withRunInIO
    $ (\ runInIO -> runInIO $ startApi configHttp)
```

Resulting error if I apply hlint suggestion

```
||     • Couldn't match type ‘App () -> IO ()’
||                      with ‘forall a. App a -> IO a’
||       Expected type: ((App () -> IO ()) -> IO ()) -> App ()
||         Actual type: ((forall a. App a -> IO a) -> IO ()) -> App ()
```

modifying the code to use brackets instead of $ (as below) prevents hlint from suggesting anything (which is desired behavior)
```
    App.runApp env $
      withRunInIO  (\runInIO ->
        runInIO $ startApi configHttp)
```

**Comment by zliu41** [comment]
What's the name of the hint (the line printed just before "Found:")?
I don't remember there being a hint that changes `.` to `$`. Doing so might create problems with impredicativity.

**Comment by JasonMFry** [comment]
> What's the name of the hint (the line printed just before "Found:")?
> I don't remember there being a hint that changes `$` to `.`. Doing so might create problems with impredicativity.

It's the `dollar` group I believe. It's recommending changing `$` to `.`.

**Comment by ndmitchell** [comment]
This is a known issue. There are lots of ways that `$` is special in Haskell (alas), including things like impredicativity. From the [HLInt manual](https://github.com/ndmitchell/hlint#bugs-and-limitations)

> Some transformed programs may require additional type signatures, particularly if the transformations trigger the monomorphism restriction or involve rank-2 types.

Also, the `dollar` group is not the default, so has slightly slower standards required from it, as only users who opt in to it see it.

**Comment by JasonMFry** [comment]
Ok, just wanted to make sure y’all were aware. I thought you would be though :)

And thanks for such a great tool!
